scripts/core/editor3/components/Editor3.jsx

```diff
@@ -20,146 +21,137 @@ import Toolbar from './toolbar';
-        this.getDecorators = this.getDecorators.bind(this);
     }
 
     /**
      * @ngdoc method
-     * @name Editor3#getDecorators
-     * @returns {Object} CompositeDecorator
-     * @description Returns the CompositeDecorator that contains the editor's decorators.
-     * It should return an array containing all of the decorators used by sub-components
-     * such as the toolbar, spellcheckers, etc.
+     * @name Editor3#focus
+     * @description Handle the editor get focus event
      */
-    getDecorators() {
-        return new CompositeDecorator(Toolbar.getDecorators());
-    }
-
-    /** Handle the editor get focus event */
     focus() {
         this.refs.editor.focus();
     }
 
-    /** Handle the editor state has been changed event*/
-    onChange(editorState, focus = false) {
-        this.setState({editorState}, () => {
-            if (focus) {
-                this.focus();
-            }
-        });
-
-        this.props.onChange(stateToHTML(editorState.getCurrentContent()));
-    }
-
-    /** Handle the editor key pressed event */
-    handleKeyCommand(command) {
-        const {editorState} = this.state;
-        const newState = RichUtils.handleKeyCommand(editorState, command);
-
-        if (newState) {
-            this.onChange(newState);
-            return true;
-        }
-        return false;
-    }
-
-    /** Handle the editor tab key pressed event */
-    onTab(e) {
-        const maxDepth = 4;
-
-        this.onChange(RichUtils.onTab(e, this.state.editorState, maxDepth));
-    }
-
+    /**
+     * @ngdoc method
+     * @name Editor3#componentWillUpdate
+     * @description Called before update, init the current editor rectangle
+     */
     componentWillUpdate() {
-        const node = ReactDOM.findDOMNode(this);
-
-        this.offset = $(node).offset();
+        this.editorRect = ReactDOM.findDOMNode(this.refs.editor).getBoundingClientRect();
     }
 
-    /** Render the editor based on current state */
+    /**
+     * @ngdoc method
+     * @name Editor3#render
+     * @param {String} command
+     * @description Render the editor based on current state
+     */
     render() {
-        const {editorState} = this.state;
-        const {showToolbar, readOnly, editorFormat} = this.props;
-
-        // If the user changes block type before entering any text, we can
-        // either style the placeholder or hide it. Let's just hide it now.
-        let className = 'Editor3-editor';
-        var contentState = editorState.getCurrentContent();
-
-        if (!contentState.hasText()) {
-            if (contentState.getBlockMap()
-                .first()
-                .getType() !== 'unstyled') {
-                className += ' Editor3-hidePlaceholder';
-            }
-        }
+        const {
+            readOnly,
+            showToolbar,
+            editorState,
+            onChange,
+            onTab,
+            handleKeyCommand
+        } = this.props;
 
         const editor = () =>
```

Comment:
I think we can refactor this whole `render()` function to use only JSX:
```js
render() {
    const {
        readOnly,
        showToolbar,
        editorState,
        onChange,
        onTab,
        handleKeyCommand
    } = this.props;

    var cx = classNames({
        'editor-root': true,
        'editor-single-line': !showToolbar
    });

    return (
        <div className={cx} onClick={this.focus}>
            {showToolbar ? <Toolbar editorOffset={this.editorRect} /> : null}
            <SpellcheckerContextMenu editorRect={this.editorRect} />
            <Editor
                editorState={editorState}
                handleKeyCommand={handleKeyCommand}
                onChange={onChange}
                onTab={onTab}
                readOnly={readOnly}
                ref="editor"
            />
        </div>
    );
}
```
`showToolbar` should not be the indicator of whether the editor is single-line or not. Instead we should have a prop that says `singleLine={true}` in that case.

Also, the classes need to be corrected to reflect this layout.